File: source/rolx-client/src/app/projects/core/phase.ts
import { Duration, TransformAsDuration } from '@app/core/util/duration';
import { TransformAsIsoDate } from '@app/core/util/iso-date';
import { assertDefined } from '@app/core/util/utils';
import * as moment from 'moment';

export class Phase {
  id!: number;
  number!: number;
  name!: string;
  fullName!: string;

  @TransformAsIsoDate()
  startDate!: moment.Moment;

  @TransformAsIsoDate()
  endDate?: moment.Moment;

  isBillable!: boolean;

  @TransformAsDuration()
  budget!: Duration;

  validateModel(): void {
    assertDefined(this, 'id');
    assertDefined(this, 'number');
    assertDefined(this, 'name');
    assertDefined(this, 'fullName');
    assertDefined(this, 'startDate');
    assertDefined(this, 'isBillable');
    assertDefined(this, 'budget');
  }

  isOpenAt(date: moment.Moment): boolean {
    return date >= this.startDate && (this.endDate == null || date <= this.endDate);
  }
}

File: source/RolXServer/RolXServer/Projects/DataAccess/Phase.cs
// -----------------------------------------------------------------------
// <copyright file="Phase.cs" company="Christian Ewald">
// Copyright (c) Christian Ewald. All rights reserved.
// Licensed under the MIT license.
// See LICENSE.md in the project root for full license information.
// </copyright>
// -----------------------------------------------------------------------

using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace RolXServer.Projects.DataAccess
{
    /// <summary>
    /// A phase of a project.
    /// </summary>
    public sealed class Phase
    {
        /// <summary>
        /// Gets or sets the identifier.
        /// </summary>
        public int Id { get; set; }

        /// <summary>
        /// Gets or sets the number.
        /// </summary>
        public int Number { get; set; }

        /// <summary>
        /// Gets or sets the project identifier.
        /// </summary>
        public int ProjectId { get; set; }

        /// <summary>
        /// Gets or sets the name.
        /// </summary>
        public string Name { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the full name.
        /// </summary>
        public string FullName { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the start date.
        /// </summary>
        [Column(TypeName = "date")]
        public DateTime StartDate { get; set; }

        /// <summary>
        /// Gets or sets the end date.
        /// </summary>
        [Column(TypeName = "date")]
        public DateTime? EndDate { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether this instance is billable.
        /// </summary>
        public bool IsBillable { get; set; }

        /// <summary>
        /// Gets or sets the time budget in seconds.
        /// </summary>
        public TimeSpan? Budget { get; set; }

        /// <summary>
        /// Gets or sets the project.
        /// </summary>
        public Project? Project { get; set; }
    }
}
